### DISCUSSION TITLE
how to upgrade the version of a plugin?

### DISCUSSION THREAD

**Discussion by mark-hahn** [vuejs/language-tools#4459] [Q&A]
VSCode says that I need to upgrade my language-plugin-pug.  I did a bunch of googling and everyone seems to think plugins and extensions are the same thing and they all say to go into the extensions even though the question uses the word plugin.  I looked and it isn't in extensions so I don't know how to proceed.   Can someone help?

**Comment by so1ve** [comment]
It is not a vscode extension. Usage here: https://github.com/vuejs/language-tools/tree/master/packages/language-plugin-pug

**Reply by mark-hahn** [reply]
I know it isn't.  I'm asking how to upgrade a plugin, not an extension.  The repo says to use NPM, but in what directory?

**Reply by johnsoncodehk** [reply]
Check `package.json` files in your project, did you find `@vue/language-plugin-pug` dependency?

**Comment by mark-hahn** [comment]
Duh.  Yes it is there and I now understand. Thanks.

I had the idea that it was a plugin to vscode not a project.  Extensions
are not limited to one project so I thought the same was true of plugins.

In my humble opinion it should be called a package, not a plugin.  What is
the difference between a plugin and a package?

**Reply by so1ve** [reply]
To be clear: this "plugin" is indeed a plugin (`VueLanguagePlugin`) that vue language tools internally invokes it to call some hooks to extend its functionality, such as pug support, once you include it in tsconfig. It is a package too, since it is published to NPM to make it possible for users to install. To make the development environment consistent between different developers, the plugin is needed to be install per project, thus both the vscode Vue extension and vue-tsc can make use of it.

**Comment by mark-hahn** [comment]
Thanks for taking the trouble to explain it.  I would never have figured
any of this out.

Feature request:  In the repo where installation instructions show the npm
command, mention that it needs to be installed per project.
